pkg/collector/process.go

```diff
@@ -0,0 +1,266 @@
+		return nil, fmt.Errorf("Error in decoding message: %v", err)
+	}
+	if message.Version != uint16(10) {
+		return nil, fmt.Errorf("Collector only supports IPFIX (v10). Invalid version %d received.", message.Version)
+	}
+	if setID == config.TemplateSetID {
+		record, err := cp.decodeTemplateSet(packetBuffer, message.ObsDomainID)
+		if err != nil {
+			return nil, fmt.Errorf("Error in decoding message: %v", err)
+		}
+		message.Record = record
+	} else {
+		record, err := cp.decodeDataSet(packetBuffer, message.ObsDomainID, setID)
+		if err != nil {
+			return nil, fmt.Errorf("Error in decoding message: %v", err)
+		}
+		message.Record = record
+	}
+	cp.messages = append(cp.messages, &message)
+	return &message, nil
+}
+
+func (cp *collectingProcess) decodeTemplateSet(templateBuffer *bytes.Buffer, obsDomainID uint32) (interface{}, error) {
+	var templateID uint16
+	var fieldCount uint16
+	err := decode(templateBuffer, &templateID, &fieldCount)
+	if err != nil {
+		return nil, fmt.Errorf("Error in decoding message: %v", err)
+	}
+	elements := make([]*templateField, 0)
+	templateMsg := entities.NewTemplateSet()
+
+	for i := 0; i < int(fieldCount); i++ {
+		element := templateField{}
+		// check whether enterprise ID is 0 or not
+		elementID := make([]byte, 2)
+		var elementLength uint16
+		err = decode(templateBuffer, &elementID, &elementLength)
+		if err != nil {
+			return nil, fmt.Errorf("Error in decoding message: %v", err)
+		}
+		indicator := elementID[0] >> 7
+		if indicator != 1 {
+			element.elementID = binary.BigEndian.Uint16(elementID)
+			element.enterpriseID = uint32(0)
+			element.elementLength = elementLength
+		} else {
+			err = decode(templateBuffer, &element.enterpriseID)
+			if err != nil {
+				return nil, fmt.Errorf("Error in decoding message: %v", err)
+			}
+			elementID[0] = elementID[0] ^ 0x80
+			element.elementID = binary.BigEndian.Uint16(elementID)
+			element.elementLength = elementLength
+		}
+		templateMsg.AddInfoElement(element.enterpriseID, element.elementID)
+		elements = append(elements, &element)
+	}
+	cp.addTemplate(obsDomainID, templateID, elements)
+	return templateMsg, nil
+}
+
+func (cp *collectingProcess) decodeDataSet(dataBuffer *bytes.Buffer, obsDomainID uint32, templateID uint16) (interface{}, error) {
+	// make sure template exists
+	template, err := cp.getTemplate(obsDomainID, templateID)
+	if err != nil {
+		return nil, fmt.Errorf("Template %d with obsDomainID %d does not exist", templateID, obsDomainID)
+	}
+	dataMsg := entities.NewDataSet()
+	for _, field := range template {
+		var length int
+		if field.elementLength == entities.VariableLength { // string
+			length = getFieldLength(dataBuffer)
+		} else {
+			length = int(field.elementLength)
+		}
+		val := dataBuffer.Next(length)
+		dataMsg.AddInfoElement(field.enterpriseID, field.elementID, val)
+	}
+	return dataMsg, nil
+}
+
+func (cp *collectingProcess) addTemplate(obsDomainID uint32, templateID uint16, elements []*templateField) {
```

Comment:
Here as well instead of templateField, may be we could use entities.infoElements. Any reason we need new struct?

Reply:
Same as above. We don't need the struct any more. Thanks!